server/remote_cache/action_cache_server/action_cache_server.go

```diff
@@ -203,19 +160,67 @@ func (s *ActionCacheServer) GetActionResult(ctx context.Context, req *repb.GetAc
 	if *checkClientActionResultDigests && req.GetCachedActionResultDigest().GetHash() != "" {
 		d, err := digest.ComputeForMessage(rsp, req.GetDigestFunction())
 		if err != nil {
-			return nil, err
+			return nil, false, 0, err
 		}
+
+		// NOTE: To avoid double-counting AC hits, callers that specify a
+		// cached_action_result_digest don't do hit tracking on their own.  This
+		// means we need to track the full response size here instead.
+		originalResultSize := int64(proto.Size(rsp))
+		// Now that we've tracked size, wipe out the response.
 		if proto.Equal(req.GetCachedActionResultDigest(), d) {
 			rsp = &repb.ActionResult{
 				ActionResultDigest: d,
 			}
 		}
+		return rsp, true, originalResultSize, nil
+	}
+
+	return rsp, true, int64(proto.Size(rsp)), nil
+}
+
+// Retrieve a cached execution result.
+//
+// Implementations SHOULD ensure that any blobs referenced from the
+// [ContentAddressableStorage][build.bazel.remote.execution.v2.ContentAddressableStorage]
+// are available at the time of returning the
+// [ActionResult][build.bazel.remote.execution.v2.ActionResult] and will be
+// for some period of time afterwards. The TTLs of the referenced blobs SHOULD be increased
+// if necessary and applicable.
+//
+// Errors:
+//
+// * `NOT_FOUND`: The requested `ActionResult` is not in the cache.
+func (s *ActionCacheServer) GetActionResult(ctx context.Context, req *repb.GetActionResultRequest) (*repb.ActionResult, error) {
+	if req.ActionDigest == nil {
+		return nil, status.InvalidArgumentError("ActionDigest is a required field")
+	}
+	rn := digest.NewACResourceName(req.GetActionDigest(), req.GetInstanceName(), req.GetDigestFunction())
+	if err := rn.Validate(); err != nil {
+		return nil, err
+	}
+	ctx, err := prefix.AttachUserPrefixToContext(ctx, s.env)
+	if err != nil {
+		return nil, err
 	}
 
-	isCacheHit = true
-	uncompressedResultSizeBytes = int64(proto.Size(rsp))
+	ht := s.env.GetHitTrackerFactory().NewACHitTracker(ctx, bazel_request.GetRequestMetadata(ctx))
+	// Fetch the "ActionResult" object which enumerates all the files in the action.
+	d := req.GetActionDigest()
+	downloadTracker := ht.TrackDownload(d)
 
-	return rsp, nil
+	rsp, isCacheHit, downloadSizeBytes, err := s.fetchActionResult(ctx, rn, req)
+	ht.SetExecutedActionMetadata(rsp.GetExecutionMetadata())
+	if isCacheHit {
+		if trackerErr := downloadTracker.CloseWithBytesTransferred(downloadSizeBytes, downloadSizeBytes, repb.Compressor_IDENTITY, "ac_server"); err != nil {
+			log.Debugf("GetActionResult: download tracker error: %s", trackerErr)
+		}
+	} else {
+		if trackerErr := ht.TrackMiss(d); err != nil {
+			log.Debugf("GetActionResult: hit tracker error: %s", trackerErr)
+		}
+	}
+	return rsp, err
```

Comment:
typo: need `trackerErr != nil` in both checks, instead of `err != nil`

Also, it looks like we could maybe simplify a bit: seems like `isCacheHit == true` if and only if `err == nil`, so `isCacheHit` could be removed.

(tangent: we probably should track an "error" instead of a "miss" if the error is something other than NotFound, but that's a pre-existing issue)

Reply:
agh, thanks for catching this.  i kinda wish this tripped a linter / nogo error, but maybe that's asking too much :P